CE-429 - Transparent Proxy for VMs (#2082)
hashicorp/web-unified-docs: content/consul/v1.22.x/content/docs/connect/proxy/transparent-proxy/vm.mdx

## Requirements

Your network must meet the following environment and software requirements to use transparent proxy.

* A running Consul 1.11.1+ datacenter with at least two Linux client nodes.
* A dedicated user to run Consul. Consul uses the user ID to exclude Consul-related traffic from transparent proxy. In this page we use the `consul` Linux user.
* [DNS forwarding](/consul/docs/manage/dns/forwarding) enabled on the Consul client nodes.
* Envoy installed on the Consul client nodes. Check [Envoy proxy configuration reference](/consul/docs/reference/proxy/envoy#envoy-and-consul-client-agent) for information over supported versions.
* A dedicated user to run Envoy. Consul uses the user ID to exclude proxy-related traffic from transparent proxy. In this page we use the `envoy` Linux user.

## Manual configuration

To manually configure Consul service mesh to use transparent proxies follow these steps:

1. [Register the service](#register-the-service)
1. [Redirect traffic to proxies](#redirect-traffic-to-proxies)
1. [Connect services](#connect-services)

### Register the service

> Configure a service to use transparent proxy mode.

Reviewer: ```suggestion
Configure a service to use transparent proxy mode in the [`connect` block of a service definition](/consul/docs/reference/service#connect).
```
To explicitly draw out the link to the reference material.

### Redirect traffic to proxies

Use the consul [`connect redirect-traffic`](/consul/commands/connect/redirect-traffic) command to enable transparent proxy on the VM for the named service.

The command includes the user ID for `consul` and `envoy` in order for transparent proxy to exclude traffic from Consul and Envoy from redirection.

```shell-session
$ consul connect redirect-traffic \
   -proxy-id="${service_name}-sidecar-proxy" \
   -proxy-uid="$(id --user envoy)" \
   -exclude-uid="$(id --user consul)"
```

<Warning>
  The command adds iptables rules that will deny existing inbound traffic unless you have specified exclusions. As a result, any connection to the VM not related to Consul and Envoy will be disconnected, including SSH.
</Warning>

To exclude inbound or outbound traffic from specific ports or IP addresses, append additional arguments specifying exclusions. 

For example, you may need to exclude inbound traffic to the SSH port for login:

```shell-session
$ consul connect redirect-traffic \
   -proxy-id="${service_name}-sidecar-proxy" \
   -proxy-uid="$(id --user envoy)" \
   -exclude-uid="$(id --user consul)" \
   -exclude-inbound-port=22
```

> If you have a mesh configuration entry defined with [`TransparentProxy.MeshDestinationsOnly=true`](/consul/docs/reference/config-entry/mesh#mesh-destinations-only), this setting will deny all inbound and outbound requests to the VM once you redirect traffic. Make sure you install the application and its dependencies before redirecting traffic since transparent proxy will deny requests to download packages by default.

Reviewer: ```suggestion
If you have a [mesh configuration entry defined with `TransparentProxy.MeshDestinationsOnly=true`](/consul/docs/reference/config-entry/mesh#mesh-destinations-only), the transparent proxy will deny all inbound and outbound requests to the VM after you redirect traffic. Because the transparent proxy denies requests to download packages, make sure you install your application and its dependencies before redirecting traffic.
```

### Connect services

If ACLs are enabled, Consul service mesh denies all traffic between services by default.

Consul configures the proxies to route traffic to the appropriate upstream services based on [service intentions](/consul/docs/reference/config-entry/service-intentions).

This example shows an intention that allows traffic between service `web` and service `api`.

```hcl
Kind = "service-intentions"
Name = "api"
Sources = [
  {
    Name   = "web"
    Action = "allow"
    Namespace  = "default"
    Partition  = "default"  
  }
]
```

> Verify you can resolve the `api` service from the `web` service client node. 

Reviewer: ```suggestion
To verify that your service is correctly configured, resolve the `api` service from the `web` service client node using Consul DNS.
```
